### PR TITLE
Test/event parser

### PR DESCRIPTION
**Pull request by TxCorpi0x**
## Description

The IBC Event parser is responsible for parsing and converting the IBC events into IBC messages understandable for the relayer chain processor. In order to increate the reliability, test cases wit sample data are added to the substrate package.

## Added Tests

- [x] `TestParsePacket`
- [x] `TestParseClient`
- [x] `TestParseClientUpdate`
- [x] `TestParseChannel`
- [x] `TestParseConnection`
- [x] `TestParseEvents`

## Fixed Bugs

- [x] Wrong casting mechanism for `uint64` and `uint32`
- [x] Make `header` of the `ClientUpdateInfo` as optional property
- [x] Timestamp parsing function

### CODE REVIEW

File: relayer/chains/substrate/event_parser.go [hunk 1]
```diff
@@ -2,16 +2,19 @@ package substrate
...
 	beefyclienttypes "github.com/ComposableFi/ics11-beefy/types"
 	clienttypes "github.com/cosmos/ibc-go/v5/modules/core/02-client/types"
+	"github.com/cosmos/ibc-go/v5/modules/core/exported"
 	"github.com/cosmos/relayer/v2/relayer/provider"
+	"github.com/spf13/cast"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
 
+// TODO: implement tests for event parser methods.
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: Looks like this TODO is no longer needed

File: relayer/chains/substrate/event_parser.go [hunk 2]
```diff
@@ -24,7 +27,7 @@ type ibcMessageInfo interface {
 }
 
 // alias for the interface map
-type ibcEventQueryItem (map[string]interface{})
+type ibcEventQueryItem = map[string]interface{}
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: Could use the alias `any` vs. `interface{}`

File: relayer/chains/substrate/event_parser.go [hunk 3]
```diff
@@ -63,6 +68,7 @@ func (scp *SubstrateChainProcessor) ibcMessagesFromEvents(
 	// add all of accumulated packets to messages
 	for _, pkt := range packetAccumulator {
 		messages = append(messages, ibcMessage{
+			// TODO (Question): will the packet type always be a receive packet?
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: I wouldn't think so but I don't think I fully understand the packet accumulator abstraction. I think @agouin would be the most insightful here.

File: relayer/chains/substrate/event_parser.go [hunk 4]
```diff
@@ -108,26 +113,27 @@ func (scp *SubstrateChainProcessor) parseEvent(
 			eventType = intoIBCEventType(eType)
 
 		case ReceivePacket, WriteAcknowledgement:
-
+			// TODO (QUESTION): the cosmos parser uses the packet accumulator for other events like the SendPacket,
+			//  is there a reason that doesn't work for us? Can receive packet
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: I'm gonna defer this question to @agouin as well

File: relayer/chains/substrate/event_parser.go [hunk 5]
```diff
@@ -145,7 +151,7 @@ func (scp *SubstrateChainProcessor) parseEvent(
 // returns the unique key for packet accumulator cache
 func genAccumKey(data interface{}) ibcPacketKey {
 	return ibcPacketKey{
-		sequence:   data.(ibcEventQueryItem)["sequence"].(uint64),
+		sequence:   cast.ToUint64(data.(ibcEventQueryItem)["sequence"].(float64)),
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: It looks like we are using type assertions to go from a string to a float64 and then casting to a uint64, this feels a little clunky.

I believe packet sequence numbers are of type `uint64` in IBC. Is there any reason why we can't handle this value as a string or at least parse/type assert right into a `uint64`?

File: relayer/chains/substrate/event_parser.go [hunk 6]
```diff
@@ -155,6 +161,7 @@ func genAccumKey(data interface{}) ibcPacketKey {
 
 // client info attributes and methods
 type clientInfo struct {
+	// TODO: is there a reason the fields are exported? Since the clientInfo struct is not exported.
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: In Go it is technically valid to have an unexported type with exported fields. You wouldn't be able to initialize a variable of type `clientInfo` from another package but you could have an exported function that returns type `clientInfo` and then the exported fields are accessible. 

If the fields are never being accessed outside of this package though I think it makes sense to make them unexported. 

See: https://www.ardanlabs.com/blog/2014/03/exportedunexported-identifiers-in-go.html

File: relayer/chains/substrate/event_parser.go [hunk 7]
```diff
@@ -190,7 +197,13 @@ func (cl *clientInfo) parseAttrs(log *zap.Logger, attributes interface{}) {
 	}
 
 	cl.ClientID = attrs["client_id"].(string)
-	cl.ClientType = attrs["client_type"].(uint32)
+
+	if cl.ClientType, err = cast.ToUint32E(attrs["client_type"].(string)); err != nil {
+		log.Error("error parsing client type: ",
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: Small nit but logs should begin with an uppercase letter and should not end with punctuation. 

We have a logging style guide which can be found here https://github.com/cosmos/relayer/blob/main/docs/logging.md

File: relayer/chains/substrate/event_parser.go [hunk 8]
```diff
@@ -244,26 +259,53 @@ func (pkt *packetInfo) MarshalLogObject(enc zapcore.ObjectEncoder) error {
...
+	var height exported.Height
+	if height, err = parseHeight(attrs["height"]); err != nil {
+		log.Error("error parsing connection height: ",
+			zap.Error(err),
+		)
+		return
+	}
+	pkt.Height = height.GetRevisionHeight()
+
+	pkt.Sequence = cast.ToUint64(attrs["sequence"].(float64))
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go:265]: See other comment regarding type assertion to `float64` and then casting to `uint64`

File: relayer/chains/substrate/event_parser.go [hunk 9]
```diff
@@ -244,26 +259,53 @@ func (pkt *packetInfo) MarshalLogObject(enc zapcore.ObjectEncoder) error {
...
 	pkt.SourcePort = attrs["source_port"].(string)
 	pkt.SourceChannel = attrs["source_channel"].(string)
 	pkt.DestPort = attrs["destination_port"].(string)
 	pkt.DestChannel = attrs["destination_channel"].(string)
-	pkt.Data = attrs["data"].([]byte)
 
-	var err error
+	if pkt.Data, err = rpcclienttypes.HexDecodeString(attrs["data"].(string)); err != nil {
+		log.Error("error parsing packet data: ",
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: Logs should begin with a capital letter and not end in punctuation.

See https://github.com/cosmos/relayer/blob/main/docs/logging.md

File: relayer/chains/substrate/event_parser.go [hunk 10]
```diff
@@ -244,26 +259,53 @@ func (pkt *packetInfo) MarshalLogObject(enc zapcore.ObjectEncoder) error {
...
 	if pkt.TimeoutHeight, err = parseHeight(attrs["timeout_height"]); err != nil {
 		log.Error("error parsing packet height: ",
 			zap.Error(err),
 		)
 		return
 	}
 
-	pkt.TimeoutTimestamp = attrs["timeout_timestamp"].(uint64)
+	if pkt.TimeoutTimestamp, err = parseTimestamp(attrs["timeout_timestamp"]); err != nil {
+		log.Error("error parsing packet timeout timestamp: ",
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: Logs should begin with a capital letter and not end in punctuation.

See https://github.com/cosmos/relayer/blob/main/docs/logging.md

File: relayer/chains/substrate/event_parser.go [hunk 11]
```diff
@@ -244,26 +259,53 @@ func (pkt *packetInfo) MarshalLogObject(enc zapcore.ObjectEncoder) error {
...
+	}
+
+	ack, ackFound := attrs["ack"]
+	if ackFound {
 
-	ack, found := attrs["ack"]
-	if found {
-		pkt.Ack = ack.([]byte)
+		if pkt.Ack, err = rpcclienttypes.HexDecodeString(ack.(string)); err != nil {
+			log.Error("error parsing ack data: ",
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: Logs should begin with a capital letter and not end in punctuation.

See https://github.com/cosmos/relayer/blob/main/docs/logging.md

File: relayer/chains/substrate/event_parser.go [hunk 12]
```diff
@@ -289,6 +331,17 @@ func (ch *channelInfo) MarshalLogObject(enc zapcore.ObjectEncoder) error {
 func (ch *channelInfo) parseAttrs(log *zap.Logger, attributes interface{}) {
 	attrs := attributes.(ibcEventQueryItem)
 
+	var err error
+
+	var height exported.Height
+	if height, err = parseHeight(attrs["height"]); err != nil {
+		log.Error("error parsing channel height: ",
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: Logs should begin with a capital letter and not end in punctuation.

See https://github.com/cosmos/relayer/blob/main/docs/logging.md

File: relayer/chains/substrate/event_parser.go [hunk 13]
```diff
@@ -312,22 +365,33 @@ func (con *connectionInfo) MarshalLogObject(enc zapcore.ObjectEncoder) error {
 func (con *connectionInfo) parseAttrs(log *zap.Logger, attributes interface{}) {
 	attrs := attributes.(ibcEventQueryItem)
 
+	var err error
+
+	var height exported.Height
+	if height, err = parseHeight(attrs["height"]); err != nil {
+		log.Error("error parsing connection height: ",
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: Logs should begin with a capital letter and not end in punctuation.

See https://github.com/cosmos/relayer/blob/main/docs/logging.md

File: relayer/chains/substrate/event_parser.go [hunk 14]
```diff
@@ -312,22 +365,33 @@ func (con *connectionInfo) MarshalLogObject(enc zapcore.ObjectEncoder) error {
...
 	con.CounterpartyConnID = attrs["counterparty_connection_id"].(string)
 }
 
-// parses the attributes of ibc core client type
+// parses the height from an input interface
 func parseHeight(i interface{}) (clienttypes.Height, error) {
 	height := i.(ibcEventQueryItem)
 
-	revisionNumber, err := strconv.ParseUint(height["revision_number"].(string), 10, 64)
+	revisionNumber, err := cast.ToUint64E(height["revision_number"].(float64))
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go:381]: Please see other comment regarding type assertion to `float64` and then casting to `uint64`

File: relayer/chains/substrate/event_parser.go [hunk 15]
```diff
@@ -312,22 +365,33 @@ func (con *connectionInfo) MarshalLogObject(enc zapcore.ObjectEncoder) error {
...
 	if err != nil {
 		return clienttypes.Height{}, fmt.Errorf("error parsing revision number: %s", err)
 	}
 
-	revisionHeight, err := strconv.ParseUint(height["revision_height"].(string), 10, 64)
+	revisionHeight, err := cast.ToUint64E(height["revision_height"].(float64))
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go:386]: Please see other comment regarding type assertion to `float64` and then casting to `uint64`

File: relayer/chains/substrate/event_parser.go [hunk 16]
```diff
@@ -338,6 +402,17 @@ func parseHeight(i interface{}) (clienttypes.Height, error) {
 	}, nil
 }
 
+// parses timestamp from an input interface
+func parseTimestamp(i interface{}) (uint64, error) {
```
Comment (jtieri) [inline relayer/chains/substrate/event_parser.go]: Could use type alias `any` vs. `interface{}`

**Review by jtieri — CHANGES_REQUESTED** [review]
Good start here, I've got a few questions/nits and also deferring a few of the questions in the comments to @agouin

File: relayer/chains/substrate/event_parser.go (hunk 3 above)
Comment (TxCorpi0x) [reply relayer/chains/substrate/event_parser.go]: @jtieri Let it be as is, We need to modify this part when the code is fully merged with ibctest and running servers

File: relayer/chains/substrate/event_parser.go (hunk 4 above)
Comment (TxCorpi0x) [reply relayer/chains/substrate/event_parser.go]: @jtieri Same as previous point, This is for composable team to check while testing via ibctest

File: relayer/chains/substrate/event_parser.go (hunk 5 above)
Comment (TxCorpi0x) [reply relayer/chains/substrate/event_parser.go]: The map is being converted from json package of golang, every number is being detected as float64

File: relayer/chains/substrate/event_parser.go (hunk 14 above)
Comment (TxCorpi0x) [reply relayer/chains/substrate/event_parser.go:381]: Same as previous point

File: relayer/chains/substrate/event_parser.go (hunk 15 above)
Comment (TxCorpi0x) [reply relayer/chains/substrate/event_parser.go:386]: Same as previous point

Comment (TxCorpi0x) [reply relayer/chains/substrate/event_parser.go:386]: Same as previous point

File: relayer/chains/substrate/event_parser.go (hunk 8 above)
Comment (TxCorpi0x) [reply relayer/chains/substrate/event_parser.go:265]: @jtieri Same as others

**Review by jtieri — APPROVED** [review]

**Pull request merged**